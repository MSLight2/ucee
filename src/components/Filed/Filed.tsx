import React, { useCallback, useEffect, useRef, useState } from 'react';
import classnames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import { transformSize } from '../../utils';

export type InputType = 'button' |
  'checkbox' |
  'color' |
  'date' |
  'datetime-local' |
  'email' |
  'file' |
  'hidden' |
  'image' |
  'month' |
  'number' |
  'password' |
  'radio' |
  'range' |
  'reset' |
  'search' |
  'submit' |
  'tel' |
  'text' |
  'time' |
  'url' |
  'week'

export interface BaseFieldProps {
    type?: InputType
    formatType?: 'integer' | 'float'
    precision?: number
    size?: 'default' | 'large'
    label?: string
    border?: boolean
    clearable?: boolean
    readonly?: boolean
    titleWidth?: string | number
    inputAlign?: 'left' | 'right'
    leftIcon?: React.ReactNode
    rightIcon?: React.ReactNode
    rightContent?: React.ReactNode
    className?: string
    // formatter?: (value: number | string | undefined) => string
    // parser?: (value: number | string | undefined) => string | number
    onChange?: React.ChangeEventHandler<HTMLInputElement>
    onClear?: Function
    onBlur?: React.FocusEventHandler<HTMLInputElement>
    onInputClick?: React.MouseEventHandler<HTMLInputElement>
    onRightClick?: React.MouseEventHandler<HTMLDivElement>
}

export type OmitAttrs = 'size' | 'type' | 'defaultValue' | 'readOnly'

export type FieldProps = Partial<
    BaseFieldProps &
    Omit<React.InputHTMLAttributes<HTMLInputElement>, OmitAttrs>
  >

const Field: React.FC<FieldProps> = (props) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [inputVal, setInputVal] = useState<any>('');
  const [isShowClear, setIsShowClear] = useState<boolean>(false);

  const {
    type,
    formatType,
    precision,
    size = 'default',
    value,
    label,
    border,
    clearable,
    readonly,
    disabled,
    titleWidth,
    inputAlign,
    leftIcon,
    rightIcon,
    rightContent,
    className,
    // formatter,
    // parser,
    onChange,
    onClear,
    onBlur,
    onInputClick,
    onRightClick,
    ...rest
  } = props;
  const baseClass = 'ucee-field';
  const fieldClassNames = classnames(
    baseClass,
    {
      [`${baseClass}--border`]: border,
      [`${baseClass}--large`]: size === 'large',
      [`${baseClass}--disabled`]: disabled,
      [`${baseClass}--right`]: inputAlign === 'right'
    },
    className
  );

  const lableProps = {
    className: `${baseClass}__label`
  };
  if (titleWidth) {
    Object.assign(lableProps, {
      style: {
        width: transformSize(titleWidth)
      }
    })
  }
  const inputProps = {
    ...rest,
    type,
    readOnly: readonly,
    className: `${baseClass}__control`,
    onClick: onInputClick
  };

  useEffect(() => {
    if (value) {
      setIsShowClear(true);
      setInputVal(value);
    }
    // eslint-disable-next-line
  }, []);

  const RenderLeftIcon = useCallback(() => {
    return (
      <>
        {leftIcon && 
          <div className={`${baseClass}__left-icon`}>{leftIcon}</div>
        }
      </>
    );
  }, [leftIcon]);
  
  const RenderRightIcon = useCallback(() => {
    return (
      <>
        {rightIcon && !clearable &&
          <div
            className={`${baseClass}__right-icon`}
            onClick={onRightClick}
          >
            {rightIcon}
          </div>
        }
      </>
    );
  }, [rightIcon, clearable, onRightClick]);

  const RenderRightContent = useCallback(() => {
    return (
      <>
        {rightContent &&
          <div className={`${baseClass}__right-content`}>
            {rightContent}
          </div>
        }
      </>
    );
  }, [rightContent]);

  const RenderClearIcon = useCallback(() => {
    return (
      <>
        {clearable && isShowClear &&
          <div
            className={`${baseClass}--icon-clear`}
            onClick={onClearHandler}
          >
            <Icon name='delete-fill' />
          </div>
        }
      </>
    );
    // eslint-disable-next-line
  }, [clearable, isShowClear, onClear]);

  function getInteger(str: string | undefined): string {
    str = str ?? '';
    str = str.replace(/\D*/g, '');
    return str;
  }
  function getFloat(str: string | undefined, decimalNum: number = 2): string {
    const decimalCountRegx = '\\d'.padEnd(2 * decimalNum, '\\d');
    // eg: /^(\d+)\.(\d\d).*$/
    const decimalRegx = `^(\\d+)\\.(${decimalCountRegx}).*$`;
    str = str ?? '';
    str = str
      .replace(/[^0-9.]*/g, '')
      .replace(/^\./, '')
      .replace(/^0{2,}/g, '0') // 不能多个0开头
      .replace(/\.{2,}/g, '.') // 只能输入一个`.`
      .replace(new RegExp(decimalRegx),'$1.$2') // 只能输入 `x` 位小数
    return str;
  }
  // 格式化输入
  function formatTypeVal (val: string | undefined) {
    if (formatType === 'integer') {
      return getInteger(val);
    } else if (formatType === 'float') {
      return getFloat(val, precision);
    }
    return val ?? '';
  }

  function onClearHandler () {
    setInputVal('');
    setIsShowClear(false);
    onClear?.();
  }
  function onChangeHandler (e: React.ChangeEvent<HTMLInputElement>) {
    let val = e.target.value;
    if (clearable) {
      val ? setIsShowClear(true) : setIsShowClear(false);
    }
    val = formatTypeVal(val);

    setInputVal(val);
    e.target.value = val;
    onChange?.(e);
  }
  function onBlurHandler (e: React.FocusEvent<HTMLInputElement>) {
    let val = e.target.value;
    val = formatTypeVal(val);

    setInputVal(val);
    e.target.value = val;
    onBlur?.(e);
  }

  return (
    <div className={fieldClassNames}>
      <RenderLeftIcon />
      <div {...lableProps}>
        <span>{label}</span>
      </div>
      <div className={`${baseClass}__value ${baseClass}__right-clear`}>
        <input
          {...inputProps}
          ref={inputRef}
          value={inputVal}
          onChange={onChangeHandler}
          onBlur={onBlurHandler}
        />
        <RenderClearIcon />
      </div>
      <RenderRightIcon />
      <RenderRightContent />
    </div>
  )
}

export default Field;