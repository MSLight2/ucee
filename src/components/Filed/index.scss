@import '../../styles/var';
@import '../../styles/mixin';

.ucee-field{
  position: relative;
  display: flex;
  padding: $field-top-padding $field-up-padding;
  box-sizing: border-box;
  color: $text-color;
  font-size: $field-font-size;
  line-height: $field-line-height;
  background-color: $white;
  overflow: hidden;
  &__label{
    flex-shrink: 0;
    width: $field-label-width;
    margin-right: $base-margin * 3;
    text-align: left;
  }
  &__value{
    position: relative;
    flex: 1;
    vertical-align: middle;
    box-sizing: border-box;
    .ucee-field__control{
      @include inputRest();
    }
  }
  &__left-icon,
  &__right-icon,
  &__right-clear{
    .ucee-icon{
      color: $gray-5;
      display: inline-block;
      font-size: 16px;
      line-height: inherit;
    }
  }
  &__left-icon{
    margin-right: $base-margin;
  }
  &__right-clear{
    .ucee-field__control{
      padding-right: 20px;
    }
  }
  &__right-icon{
    padding: 0 8px;
    margin-right: -8px;
  }
  &__right-content{
    flex-shrink: 0;
    padding-left: 8px;
  }
  &--icon-clear{
    position: absolute;
    top: 50%;
    right: 0;
    transform: translateY(-50%);
  }
  &--border::after{
    @include hairline('bottom', $gray-3, 16px, 16px);
  }
  &--large{
    padding-top: $field-large-top-padding;
    padding-bottom: $field-large-top-padding;
  }
  &--disabled{
    color: $gray-5;
    pointer-events: none;
    user-select: none;
    .ucee-field__label{
      color: $gray-5;
    }
  }
  &--right{
    .ucee-field__control{
      text-align: right;
    }
  }
}