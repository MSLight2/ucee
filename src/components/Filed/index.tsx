import Filed from './Filed';

export type {
  InputType,
  BaseFieldProps,
  FieldProps
} from './Filed';
export default Filed;