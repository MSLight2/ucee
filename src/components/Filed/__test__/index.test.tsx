import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Filed from '..';

describe('Filed', () => {
  test('should trigger onChange', () => {
    const onChange = jest.fn();
    const { container } = render(<Filed onChange={onChange} />);
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.change(el, { target: { value: '12' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('12');
  });

  test('should trigger onClear', () => {
    const onClear = jest.fn();
    const { container } = render(<Filed clearable onClear={onClear} value='123' />);
    const iconEle = container.querySelector('.ucee-field--icon-clear') as Element;
    const el = container.querySelector('.ucee-field__control') as HTMLInputElement;

    fireEvent.click(iconEle);
    expect(onClear).toHaveBeenCalled();
    expect(el.value).toEqual('');
  });

  test('should toggle clearable', () => {
    const onChange = jest.fn();
    const { container } = render(<Filed clearable onChange={onChange} />);
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.change(el, { target: { value: '123' } });
    expect(container.querySelector('.ucee-field--icon-clear')).toBeTruthy();
    fireEvent.change(el, { target: { value: '' } });
    expect(container.querySelector('.ucee-field--icon-clear')).toBeFalsy();
  });

  test('should trigger onBlur', () => {
    const onBlur = jest.fn();
    const { container } = render(<Filed onBlur={onBlur} />);
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.blur(el);
    expect(onBlur).toHaveBeenCalled();
  });

  test('should trigger onRightClick', () => {
    const onRightClick = jest.fn();
    const { getByTestId } = render(
      <Filed
        onRightClick={onRightClick}
        rightIcon={<span data-testid="right-icon">icon</span>}
      />
    );
    const spanEle = getByTestId('right-icon');

    fireEvent.click(spanEle);
    expect(onRightClick).toHaveBeenCalled();
  });

  test('should trigger onInputClick', () => {
    const onInputClick = jest.fn();
    const { container } = render(<Filed onInputClick={onInputClick} />);
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.click(el);
    expect(onInputClick).toHaveBeenCalled();
  });

  test('should render field label', () => {
    const { container } = render(<Filed label="文本" titleWidth={200} />);
    expect(container).toMatchSnapshot();
  });

  test('should render leftIcon', () => {
    const { container } = render(<Filed leftIcon={<span>icon</span>} />);
    expect(container).toMatchSnapshot();
  });

  test('should output integer number when formatType prop is integer', () => {
    const onChange = jest.fn();
    const { container } = render(<Filed formatType='integer' onChange={onChange} />);
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.change(el, { target: { value: '123abc' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('123');
    fireEvent.change(el, { target: { value: '1a2b3c' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('123');
  });

  test('should output float number when formatType prop is float', () => {
    const onChange = jest.fn();
    const { container } = render(
      <Filed formatType='float' precision={4} onChange={onChange} />
    );
    const el = container.querySelector('.ucee-field__control') as Element;

    fireEvent.change(el, { target: { value: '..123' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('.123');
    fireEvent.change(el, { target: { value: '00.123' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('0.123');
    fireEvent.change(el, { target: { value: '1.23456' } });
    expect(onChange.mock.calls[0][0].target.value).toEqual('1.2345');
  });
});