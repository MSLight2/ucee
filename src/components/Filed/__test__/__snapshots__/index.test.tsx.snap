// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filed should render field label 1`] = `
<div>
  <div
    class="ucee-field"
  >
    <div
      class="ucee-field__label"
      style="width: 200px;"
    >
      <span>
        文本
      </span>
    </div>
    <div
      class="ucee-field__value ucee-field__right-clear"
    >
      <input
        class="ucee-field__control"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Filed should render leftIcon 1`] = `
<div>
  <div
    class="ucee-field"
  >
    <div
      class="ucee-field__left-icon"
    >
      <span>
        icon
      </span>
    </div>
    <div
      class="ucee-field__label"
    >
      <span />
    </div>
    <div
      class="ucee-field__value ucee-field__right-clear"
    >
      <input
        class="ucee-field__control"
        value=""
      />
    </div>
  </div>
</div>
`;
